From c3301e014201ddc7d04c8076ce928328925fb10d Mon Sep 17 00:00:00 2001
From: Tycho Andersen <tycho.andersen@canonical.com>
Date: Wed, 6 Jan 2016 13:26:57 -0700
Subject: export: fix crash when metadata.yaml is missing

Two parts to this fix:

1. We should compute the temporary metadata.yaml's offset path for
   inclusion in the tar file relative to the temp dir, instead of the
   container's rootfs.

2. Only try to include the metadata.yaml once if it doesn't exist (i.e.
   shift the container filepath's include code into the else branch)

Closes #1448

Signed-off-by: Tycho Andersen <tycho.andersen@canonical.com>
---
 lxd/container_lxc.go | 29 +++++++++++++++--------------
 1 file changed, 15 insertions(+), 14 deletions(-)

diff --git a/lxd/container_lxc.go b/lxd/container_lxc.go
index 6afe7ce..1d55969 100644
--- a/lxd/container_lxc.go
+++ b/lxd/container_lxc.go
@@ -1934,27 +1934,28 @@ func (c *containerLXC) Export(w io.Writer) error {
 			return err
 		}
 
-		if err := c.tarStoreFile(linkmap, offset, tw, f.Name(), fi); err != nil {
+		tmpOffset := len(path.Dir(f.Name())) + 1
+		if err := c.tarStoreFile(linkmap, tmpOffset, tw, f.Name(), fi); err != nil {
 			shared.Debugf("Error writing to tarfile: %s", err)
 			tw.Close()
 			return err
 		}
 
 		fnam = f.Name()
-	}
-
-	// Include metadata.yaml in the tarball
-	fi, err := os.Lstat(fnam)
-	if err != nil {
-		shared.Debugf("Error statting %s during export", fnam)
-		tw.Close()
-		return err
-	}
+	} else {
+		// Include metadata.yaml in the tarball
+		fi, err := os.Lstat(fnam)
+		if err != nil {
+			shared.Debugf("Error statting %s during export", fnam)
+			tw.Close()
+			return err
+		}
 
-	if err := c.tarStoreFile(linkmap, offset, tw, fnam, fi); err != nil {
-		shared.Debugf("Error writing to tarfile: %s", err)
-		tw.Close()
-		return err
+		if err := c.tarStoreFile(linkmap, offset, tw, fnam, fi); err != nil {
+			shared.Debugf("Error writing to tarfile: %s", err)
+			tw.Close()
+			return err
+		}
 	}
 
 	// Include all the rootfs files
